src/libcore/iter.rs

```diff
@@ -1565,24 +1565,13 @@ pub struct Map<I, F> {
     f: F,
 }
 
-impl<I: Iterator, F, B> Map<I, F> where F: FnMut(I::Item) -> B {
-    #[inline]
-    fn do_map(&mut self, elt: Option<I::Item>) -> Option<B> {
-        match elt {
-            Some(a) => Some((self.f)(a)),
-            _ => None
-        }
-    }
-}
-
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<B, I: Iterator, F> Iterator for Map<I, F> where F: FnMut(I::Item) -> B {
     type Item = B;
 
     #[inline]
     fn next(&mut self) -> Option<B> {
-        let next = self.iter.next();
-        self.do_map(next)
+        self.iter.next().map(|a| (self.f)(a))
```

Comment:
Due to the auto implmeting among the ```Fn, FnMut, FnOnce```, we have some constrains.

Among these 3 types: ```&mut T where T: Fn, &mut T where T:FnMut, &mut T where T:FnOnce```, we can only impl one ```Fn*``` trait to them, otherwise they will conflict. The only choice is ```impl FnMut for &mut T where T:FnMut```, other choices make no sense.

The same, Among these 3 types: ```&T where T: Fn, &T where T:FnMut, &T where T:FnOnce```, we can only impl one ```Fn*``` trait to them, otherwise they will conflict. The best choice is ```impl Fn for &T where T:Fn```, other choices will cover less situation.

Could I implement the above?